### ISSUE TITLE
Syn is great But Documentation is very Poor

### ISSUE THREAD

**Issue by douggynix** [dtolnay/syn#1533]
I have been using Syn crate. however, documentation is very poor. It takes you to navigate the library source code to get some insight about how to use and what to use.

I would propose that we think about documentation as a lot of users in forums are complaining how complicated it is to make use of this syn Lib. And with proper examples , that can give users a good insight and starting point for more complex parsing.

I would propose some scenarios for example

What would be the code to parse these below with syn:

**Example1**
#[derive(Foo)]
#[bar( prop1="one" , prop2=["two","three"]  )]
#[bar( prop1=("a", "b","c")  ]
struct FooObj

**Example2**
#[derive(Foo)]
#[bar( prop1="one" , prop2=[("two",2) ,("three",3  )]
#[bar( prop1=("a", "b","c")  ]
struct FooObj

I have seen people struggling with those simple examples. and I have not seen nowhere it is documented how to deal with them.
More intutitive documentation with real world scenarios like above would make more sense for users who need to get hands on this library.

It would be good not to have such comments like the one i saw in reddit about the library. I think if you document how to handle such scenarios, a lot of pending questions will be answered. Context is really important when it comes to documentation.
[https://www.reddit.com/r/rust/comments/j6afjj/procedural_macro_parse_macro_attributesarray_with/](https://www.reddit.com/r/rust/comments/j6afjj/procedural_macro_parse_macro_attributesarray_with/)
![image](https://github.com/dtolnay/syn/assets/11457752/7f6f1f86-e788-4980-a5b9-f90e117749ba)
![image](https://github.com/dtolnay/syn/assets/11457752/95f40051-132a-4c1c-9064-5b5c688806a7)

**Comment by dtolnay** [comment]
Your link is from over 3 years ago. A lot of examples of attribute parsing have been added since then, including in https://docs.rs/syn/2.0.39/syn/struct.Attribute.html and https://docs.rs/syn/2.0.39/syn/meta/struct.ParseNestedMeta.html.

**Comment by douggynix** [comment]
@dtolnay Thanks for your feedback on this issue. it would be great having these links on the landing page of the syn crate for users like me. 
By the way i followed your instructions in the ParseNestedMeta. For example, i would to print at least all the keys within #[mapper] attribute. Consider I have a vector of mapper attributes. and the  order of the keys for keys are randomn. And I woud like to list keys name for now. and I will leave the parsing of the value later on.
Running the code below gave me this error. I am wondering if **parse_nested_meta** is a good fit for what I want to achieve. I have been looking at delving myself in documentation how to do it and i have no clue yet how to sort it out with syn. it always stop after parsing the first key and resulted into error.

**key_name=map
Error: Error("expected `,`")**

`use std::error::Error;
use syn::{parse_quote, Attribute, LitStr, Token};

fn main() -> syn::Result<()>{
    let attr1: Attribute = parse_quote! {
        #[mapper(map=[("username:login",true)], dto="ProfileDto", include_all="true", except=["password", "age"])]
    };

    let attr2: Attribute = parse_quote! {
        #[mapper(dto="LoginDto" , map=[("username:login",true),("password",true)], look_at="this")]
    };

    let attributes = vec![attr1,attr2];

    for attr in attributes{
        attr.parse_nested_meta(|meta| {
            let key_name = meta.path
            .get_ident()
            .map(|ident| ident.to_string());

             
            if let Some(name) = key_name {
                println!("key_name={}",name);
            }
            else{
                println!("key_name is empty")
            }
            syn::Result::Ok(())
        })?;
    }

    Ok(())
};
`

**Comment by douggynix** [comment]
An update , i found a solution. But i don't think, it's that optimal in terms of lines of code needed to achieve it.
it took me multiple  iterative loops and nested parsing of Expr type with nested logic of destructing Expr enum type returned.
I use parse_args_with() method from an Attribute type to get an iterator of MetaNameValue that i iterate through dynamically to parse each each expected keyname  and value type inside my attributes.
**let nested = attr.parse_args_with(Punctuated::<Meta, Token![,]>::parse_terminated)?;**

Here below the complete code that does the job. i am wondering if i can get that done with a single parse instead of having to destructure Expr enum to its lowest type level.
`
use std::error::Error;

use syn::{parse_quote, Attribute, LitStr, Token, punctuated::Punctuated, Meta, Expr, Lit, ExprLit};


fn main() -> syn::Result<()>{
    let attr1: Attribute = parse_quote! {
        #[mapper(map=[("username:login",true)], dto="ProfileDto", include_all="true", except=["password", "age"])]
    };

    let attr2: Attribute = parse_quote! {
        #[mapper(dto="LoginDto" , map=[("username:login",true),("password",true)], look_at="this")]
    };

    let attributes = vec![attr1,attr2];

    for attr in attributes{
        let nested = attr.parse_args_with(Punctuated::<Meta, Token![,]>::parse_terminated)?;
        println!("nested count={:?}",nested.iter().count());
        nested.iter().for_each(|meta| {
            if let Meta::NameValue(metaname) = meta {
                let ident = metaname.path.get_ident().unwrap();
                let keyname = ident.to_string();
                //println!("keyname={}",keyname);

                if let Expr::Lit(expr) =  & metaname.value{
                    if let Lit::Str(lit_str) = & expr.lit{
                        println!("{}={}",keyname,lit_str.value())
                    }
                }

                if let Expr::Array(expr_arr) = & metaname.value{
                    //println!("{} array has {} elements",keyname,expr_arr.elems.iter().clone().count());

                    if keyname.eq_ignore_ascii_case("map"){
                        //map is a vec of tuples such as map=[("f1",true),("f2",false)]
                        let map_tuple = parse_array_of_tuple(expr_arr);
                        println!("{}={:?}",keyname,map_tuple);
                    }
                    else if keyname.eq_ignore_ascii_case("except"){
                        //except is a vec of string such as except=["val1","val2"]
                        let except_arr = parse_array_of_string(expr_arr);
                        println!("{}={:?}",keyname, except_arr);
                    }
                }
            }
        });
    }//end of for

    Ok(())
}

fn parse_array_of_tuple(expr_arr: &syn::ExprArray) -> Vec<(String,bool)> {
    let mut vec_tuple: Vec<(String,bool)> = Vec::new();

    for elem in expr_arr.elems.iter(){
        if let Expr::Tuple(el_exp) = elem{
            //println!("{} content  is a Tuple",keyname);
            let mut str_val : Option<String> = None;
            let mut flag: Option<bool> = None;
            for content_expr in el_exp.elems.iter(){
                if let Expr::Lit(content_lit) = content_expr{
                    if let Lit::Str(content) = & content_lit.lit{
                        //print!("valueStr={}",content.value());
                        str_val = content.value().into();
                    }

                    if let Lit::Bool(content) = & content_lit.lit{
                        //print!("  valueBool={}",content.value());
                        flag = content.value.into();
                    }
                }
            }

            if(str_val.is_some() && flag.is_some()){
                vec_tuple.push((str_val.unwrap() , flag.unwrap()));
            }   
            //println!("");
        }
    }

    return vec_tuple
}


fn parse_array_of_string(expr_arr: &syn::ExprArray) -> Vec<String> {
    let mut vec_str : Vec<String>  = Vec::new();
    for elem in expr_arr.elems.iter(){
        if let Expr::Lit(lit_expr) = elem{
            //println!("{} content  is a String",keyname);
            if let Lit::Str(content) = & lit_expr.lit{
                //print!("valueStr={}, ",content.value());
                vec_str.push(content.value());
            }

        }
    }
   return vec_str
}
`

**Comment by douggynix** [comment]
@dtolnay : we can close this thread. I had figured out with multiple attempts to find out how to solve my issue.
Thanks as well for your quick response and the links as well. For an end-user of your product, i would suggest that you provide those links from the main pages of the library in crate.io. 

And I have found real case scenario from this link below as well with working example that I inspired from in order to resolve my issues:
[https://github.com/jonhoo/proc-macro-workshop/tree/master](https://github.com/jonhoo/proc-macro-workshop/tree/master)

thank you :)
